File: week_1/code/longest_substring.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
    const charSet = new Set();
    let l = 0;
    let res = 0;

    for (let r = 0; r < s.length; r++) {
        while (charSet.has(s[r])) {
            charSet.delete(s[l]);
            l++;
        }
        charSet.add(s[r]);
        res = Math.max(res, r - l + 1);
    }
    return res;
};

// Test cases
console.log("Test 1 'abcabcbb':", lengthOfLongestSubstring("abcabcbb"), "(Expected: 3)");
console.log("Test 2 'bbbbb':", lengthOfLongestSubstring("bbbbb"), "(Expected: 1)");
console.log("Test 3 'pwwkew':", lengthOfLongestSubstring("pwwkew"), "(Expected: 3)");

File: week_1/code/longest_substring.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        """
        Given a string s, find the length of the longest substring without repeating characters.
        """
        # TODO: Implement using Sliding Window + Set
        charSet = set()
        l = 0
        res = 0

        for r in range(len(s)):
            while s[r] in charSet:
                charSet.remove(s[l])
                l += 1
            charSet.add(s[r])
            res = max(res, r - l + 1)
        
        return res

# Test cases
if __name__ == "__main__":
    sol = Solution()
    print(f"Test 1 'abcabcbb': {sol.lengthOfLongestSubstring('abcabcbb')} (Expected: 3)")
    print(f"Test 2 'bbbbb': {sol.lengthOfLongestSubstring('bbbbb')} (Expected: 1)")
    print(f"Test 3 'pwwkew': {sol.lengthOfLongestSubstring('pwwkew')} (Expected: 3)")
